### PR TITLE
refactor: add `Cow` type

### PR DESCRIPTION
**Pull request by DonIsaac**

### CODE REVIEW

File: src/util/cow.zig [hunk 1]
```diff
@@ -0,0 +1,169 @@
+const std = @import("std");
+const Allocator = std.mem.Allocator;
+
+const assert = @import("../util.zig").assert;
+const debugAssert = @import("../util.zig").debugAssert;
+const IS_DEBUG = @import("../util.zig").IS_DEBUG;
```
Comment (MichaelBelousov) [inline src/util/cow.zig:6]: nitpick: I think inlining this kind of check is worth it

File: src/util/cow.zig [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+    const MutSlice = if (sentinel) [:0]u8 else []u8;
+    const Slice = if (sentinel) [:0]const u8 else []const u8;
+
+    return struct {
+        /// Does this `Cow` own its data, or is it borrowing it from someone
+        /// else? `true` if borrowed.
+        ///
+        /// **HEY YOU**. Yeah, you. Are you using this struct externally? If so,
+        /// _DON'T MUTATE THIS FIELD_!
+        borrowed: bool,
```
Comment (MichaelBelousov) [inline src/util/cow.zig:23]: if that's true, why not prefix it with `_`?

File: src/util/cow.zig [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+const assert = @import("../util.zig").assert;
+const debugAssert = @import("../util.zig").debugAssert;
+const IS_DEBUG = @import("../util.zig").IS_DEBUG;
+
+/// A clone-on-write type monomorphized for strings.
+///
+/// I tried adding this to smart-pointers but it turned out to be a pain to
+/// generalize.  I need this structure _right now_; we should go back and
+/// move+generalize this code later.
+pub fn Cow(comptime sentinel: bool) type {
```
Comment (MichaelBelousov) [inline src/util/cow.zig:13]: weird but why not take the sentinel directly?

Comment (MichaelBelousov) [reply src/util/cow.zig:13]: I would actually suggest:
```zig
pub const Cow = CowSentinel(null);
pub const CowZ = CowSentinel(0);
```

File: src/util/cow.zig [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+        ///
+        /// Do _not_ directly mutate `str` on borrowed `Cow`s. In the best case
+        /// this may cause business-level corruption, in the worst case it will
+        /// trigger a segfalt.
+        ///
+        /// Implementation notes:
+        /// - Whether this is mutable or not depends on `borrowed`. Storing this
+        ///   as a const slice variant make the safer thing the default.
+        /// - I really, _really_ wish Zig had private container members...
+        str: Slice,
```
Comment (MichaelBelousov) [inline src/util/cow.zig:34]: why not _str?

File: src/util/cow.zig (hunk 2 above)
Comment (DonIsaac) [reply src/util/cow.zig:23]: It can be read by external consumers, just not modified

**Pull request merged**